src/main/java/hudson/tasks/junit/CaseResult.java

```diff
@@ -256,10 +262,25 @@ private static String getSkippedMessage(Element testCase) {
         return message;
     }
 
-    public String getDisplayName() {
+    public String getTransformedTestName() {
         return TestNameTransformer.getTransformedName(testName);
     }
 
+    public String getDisplayName() {
+        // Only prepend the enclosing flow node names if there are any and the run this is in has multiple blocks directly containing
+        // test results.
+        if (!getEnclosingFlowNodeNames().isEmpty()) {
```

Comment:
@svanoort re the cache you're talking about - do you basically just mean running, say,
```
LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
FlowNode enclosing = scanner.findFirstMatch(this, FlowScanningUtils.MATCH_BLOCK_START);
if (enclosing != null) {
  this.enclosingId = enclosing.getId();
}
```
...at `FlowNode` instantiation time? And then providing something like this on `FlowNode`:
```
    private transient List<FlowNode> enclosingBlocks;

    @Nonnull
    public List<FlowNode> getEnclosingBlocks() {
        if (enclosingBlocks == null) {
            enclosingBlocks = new ArrayList<>();
            if (enclosingId != null) {
                try {
                    FlowNode enclosingNode = exec.getNode(enclosingId);
                    if (enclosingNode != null) {
                        enclosingBlocks.add(enclosingNode);
                        enclosingBlocks.addAll(enclosingNode.getEnclosingBlocks());
                    }
                } catch (IOException e) {
                    LOGGER.log(Level.WARNING, "failed to load parents of " + id, e);
                }
            }
        }
        return enclosingBlocks;
    }
```